docs/ECOS.md

```diff
@@ -0,0 +1,141 @@
+
+Since nothing happens with EVE unless initiated by a Controller through a certain set of configuration directives, this section focuses on:
+
+   #1 ECO state transition diagram
+   #2 configuration directives utilized by an EVE controller to move through the states
+   #3 metrics and events communicated back by EVE to the controller after any transition is attempted
+
+### ECO Runtime State Transition Diagram
+
+We will try to align with [OCI Lifecycle](https://github.com/opencontainers/runtime-spec/blob/master/runtime.md#lifecycle) where possible, but we may need some extra states as well.
+
+### Detailed description of states
+
+Currently we support the following states for the ECOs:
+
+* ECI has been transferred to EVE's storage area
+* Resources (CPU, memory, networking, I/O) have been committed to an ECO but the container has not been started yet
+* ECO is running
+* ECO is stopped (the resources are still pinned and mutable resources such as R/W storage remain in the state they were when ECO was running)
+* ECO is restarted
+* ECO resources are purged (TBD)
+* ECO is deleted (all the resources are released)
+
+### Controller config ECO schema
+
+As part of a regular config that EVE recieves from its controller, each ECO gets configured by an entry in an array having:
+
+* displayname
+* uuidandversion
+* activate
+* purge
+* restart
+* fixedresources
+  * virtualizationMode
+    * memory
+    * maxmem
+    * vcpus
+    * rootdev
+  * interfaces
+    * name
+    * acls
+    * networkId
+  * drives
+    * drvtype
+    * target
+    * maxsizebytes
+    * readonly
+    * preserve
+    * image
+      * name
+      * uuidandversion
+        * iformat
+        * sizeBytes
+        * sha256
+        * dsId
+        * siginfo
+
+## ECO Image Format
+
+This specification defines an ECI, consisting of a manifest, a set of binary layers, and a configuration. Both binary layers and configuration are represented as opaque binary blobs and referenced in a [CAS manner](https://en.wikipedia.org/wiki/Content-addressable_storage) by a ``<algo>:<content hash value>``. This allows us to have a structure similar to a [Merkle tree](https://en.wikipedia.org/wiki/Merkle_tree) with every element in the structure (starting from leaves) presumed immutable. The root of this tree is always a configuration object that in turn references other components of the tree. Binary content of the ECI, then, becomes synomynous with the ``<algo>:<content hash value>`` handle for the configuration object and thus is presumed to be immutable and content addressable.
```

Comment:
"consisting of a manifest, a set of binary layers, and a configuration"

Make each of these into a bullet point?

Reply:
"binary layers" - how would we do that? For OCI containers, it makes sense, since it is just filesystem layers that are applied one on top of the other. How would we do that here? Aren't they just disk images?

Reply:
"Make each of these into a bullet point?" -- good idea -- done.

"how would we do that?" -- we have ideas ;-)